File: 101-200/113 Path Sum II/Solution.py
class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        result = []
        self.helper(root, sum, result, [])
        return result

    def helper(self, root: TreeNode, sum: int, result: list, temp: list):
        if root is None:
            return

        temp.append(root.val)

        if root.val == sum and root.left is None and root.right is None:
            result.append(temp.copy())

        if root.left:
            self.helper(root.left, sum - root.val, result, temp)
        if root.right:
            self.helper(root.right, sum - root.val, result, temp)
        temp.pop()

File: 101-200/113 Path Sum II/Solution.java
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> result = new ArrayList<>();
        helper(root, sum, result, new ArrayList<>());
        return result;
    }

    private void helper(TreeNode root, int sum, List<List<Integer>> result, List<Integer> temp) {
        if (root == null) return;

        temp.add(root.val);

        if (root.val == sum && root.left == null && root. right == null) {
            result.add(new ArrayList<>(temp));
        }

        helper(root.left, sum - root.val, result, temp);
        helper(root.right, sum - root.val, result, temp);
        temp.remove(temp.size() - 1);
    }
}
